File: translate.py
#/usr/bin/env python
#call baidu api to translate word
#coding=utf8

import httplib
import md5
import urllib
import random
import argparse
import json
import re

def analysis_lang():
	pattern_e = '[a-z]+'
	parser = argparse.ArgumentParser()
	parser.add_argument('word', type=str, help='Need a word to translate')
	word = parser.parse_args().word
	if re.match(pattern_e, word, re.I):
		fromLang = 'en'
		toLang = 'zh'
	else:
		fromLang = 'zh'
		toLang = 'en'
	return word, fromLang, toLang

def translate(option):
	appid = 'id'
	secretKey = 'password'
	word = option[0]
	fromLang = option[1]
	toLang = option[2]
	httpClient = None
	myurl = '/api/trans/vip/translate'
	salt = random.randint(32768, 65536)

	sign = appid+word+str(salt)+secretKey
	m1 = md5.new()
	m1.update(sign)
	sign = m1.hexdigest()
	myurl = myurl+'?appid='+appid+'&q='+urllib.quote(word)+'&from='+fromLang+'&to='+toLang+'&salt='+str(salt)+'&sign='+sign

	try:
	    httpClient = httplib.HTTPConnection('api.fanyi.baidu.com')
	    httpClient.request('GET', myurl)

	    response = httpClient.getresponse()
	    js = json.load(response)
	    return js['trans_result'][0]['dst'], word
	except Exception, e:
	    print e
	finally:
	    if httpClient:
	        httpClient.close()

if __name__ == '__main__':
	result = translate(analysis_lang())
	print result[1], '--->', result[0]

File: translate.go
package main

import (
	"crypto/md5"
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"math/rand"
	"net/http"
	"os"
	"regexp"
	"strconv"
)

func main() {
	word, err := getWord()
	if err != nil {
		fmt.Println(err.Error())
	} else {
		translate(word)
	}
}

type response struct {
	From string   `json:"from"`
	To   string   `json:"to"`
	R    []result `json:"trans_result"`
}

type result struct {
	Src string `json:"src"`
	Dst string `json:"dst"`
}

func getWord() (lang []string, err error) {
	if len(os.Args) != 2 {
		e := fmt.Errorf("Please input a word to translate!")
		return lang, e
	} else {
		word := os.Args[1]
		lang := append(lang, word)
		isok, _ := regexp.MatchString("[a-z]+", word)
		if isok {
			lang = append(lang, "en", "zh")
			return lang, nil
		} else {
			lang = append(lang, "zh", "en")
			return lang, nil
		}
	}
}

func translate(word []string) {
	var r response
	appid := "id"
	password := "password"
	salt := strconv.Itoa(rand.Int())

	str := appid + word[0] + salt + password
	m := md5.New()
	io.WriteString(m, str)
	sign := fmt.Sprintf("%x", m.Sum(nil))
	url := "/api/trans/vip/translate" + "?appid=" + appid + "&q=" + word[0] + "&from=" + word[1] + "&to=" + word[2] + "&salt=" + salt + "&sign=" + sign
	api := "http://api.fanyi.baidu.com" + url

	res, err := http.Get(api)
	defer res.Body.Close()
	if err != nil {
		fmt.Println("error of get api")
	}
	content, err := ioutil.ReadAll(res.Body)
	if err != nil {
		fmt.Println("error of read response body")
	}
	json.Unmarshal(content, &r)
	l := len(r.R)
	fmt.Println(r.R[l-1].Src, "--->", r.R[l-1].Dst)
}
